_posts/2022-09-28-import-big-csv.md

```diff
@@ -0,0 +1,306 @@
+families:
+- One is the **Dataset API** and;
+- the other is the one that I will refer to as the **Single file** API.
+
+The Dataset API is very flexible.  It can read multiple file formats,
+you can point to a folder with multiple files and create a dataset
+from them, it can read datasets from multiple sources (even combining
+remote and local sources). This API can also be used to read single
+files that are too large to fit in memory. This works because the
+files are not loaded in memory. The functions scan the content so they
+know where to look for the data and what is the schema (the data types
+and names of each column). When you query the data, there is a little
+overhead because the query engine needs to first read the data before
+it can operate on it. (If you want to see some examples of what the
+Dataset API can do, check out the previous posts).
+
+The Single file API groups a series of functions that are specific to
+each file format (CSV, JSON, Parquet, Feather/Arrow, ORC). They work
+on one file at a time, and they load the data in memory. So depending
+on the size of your file and the amount of memory you have available
+on your system, it might not be possible to load the dataset this
+way. If you can load this dataset, queries will however run faster
+because the data is already in memory and will be quick to access for
+the query engine.
+
+In this post, we will explore how to convert a large CSV in the Apache
+Parquet format using the Dataset and the Single file APIs with code
+examples in R and Python.
+
+## The Single file API in R
+
+The functions in the Single file API in R start with `read_` or
+`write_` followed by the name of the file format. For instance,
+`read_csv_arrow()`, `read_parquet()`, `read_feather()`.
+
+To read the data with our 15 GB CSV file, we would use:
+
+```r
+library(arrow)
+
+data <- read_csv_arrow(
+  "~/dataset/path_to_file.csv",
+  as_data_frame = FALSE
+)
+```
+
+Using `as_data_frame = FALSE` keeps the result as an Arrow table which
+is a better representation for a file of this size. Attempting to
+convert it into a data frame will most likely cause you to run out of
+memory. This step takes about 15s on my system.
+
+At this point, you have an Arrow formatted data to work with. You can
+stop here and do your analysis. 
+
+However, we saw in the previous post that Apache Parquet offered a
+good balance between file size and reading speed. To convert this file
+into Parquet using the Single file API, you would use:
+
+```r
+write_parquet(data, "~/dataset/data.parquet")
+```
+
+Creating this file takes about 85 seconds on my system. The resulting
+file is about 9.5 GB. You can then load it again the next time you
+need to work with it usiing:
+
+```r
+data <- read_parquet("~/dataset/data.parquet", as_data_frame = FALSE)
+```
+
+Reading this file takes about 25 seconds.
+
+A query looking at the number of occurences for each of the 10
```

Comment:
Could you insert the code for this query? Seeing it in words alone slowly lost me as I moved through the article.